isisd/isis_sr.c

```diff
@@ -0,0 +1,1654 @@
+	case CIRCUIT_T_BROADCAST:
+		ladj_sid = adj->sr.ipv4.ladj_sid;
+		if (ladj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, ladj_sid->sid, AF_INET,
+				&adj->ipv4_addresses[0]);
+			isis_tlvs_del_lan_adj_sid(circuit->ext, ladj_sid);
+			adj->sr.ipv4.ladj_sid = NULL;
+		}
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = adj->sr.ipv4.adj_sid;
+		if (adj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, adj_sid->sid, AF_INET,
+				&adj->ipv4_addresses[0]);
+			isis_tlvs_del_adj_sid(circuit->ext, adj_sid);
+			adj->sr.ipv4.adj_sid = NULL;
+		}
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+}
+
+static void isis_sr_circuit_del_ipv6_adj_sid(struct isis_circuit *circuit,
+					     struct isis_adjacency *adj)
+{
+	struct isis_adj_sid *adj_sid;
+	struct isis_lan_adj_sid *ladj_sid;
+
+	switch (circuit->circ_type) {
+	case CIRCUIT_T_BROADCAST:
+		ladj_sid = adj->sr.ipv6.ladj_sid;
+		if (ladj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, ladj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_lan_adj_sid(circuit->ext, ladj_sid);
+			adj->sr.ipv6.ladj_sid = NULL;
+		}
+		break;
+	case CIRCUIT_T_P2P:
+		adj_sid = adj->sr.ipv6.adj_sid;
+		if (adj_sid) {
+			isis_sr_adj_sid_install_uninstall(
+				false, circuit, adj, adj_sid->sid, AF_INET6,
+				&adj->ipv6_addresses[0]);
+			isis_tlvs_del_adj_sid(circuit->ext, adj_sid);
+			adj->sr.ipv6.adj_sid = NULL;
+		}
+		break;
+	default:
+		flog_err(EC_LIB_DEVELOPMENT, "%s: unexpected circuit type: %u",
+			 __func__, circuit->circ_type);
+		exit(1);
+	}
+}
+
+#ifndef FABRICD
+static int isis_sr_adj_state_change(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	if (adj->adj_state != ISIS_ADJ_UP) {
+		isis_sr_circuit_del_ipv4_adj_sid(circuit, adj);
+		isis_sr_circuit_del_ipv6_adj_sid(circuit, adj);
+	}
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv4_enabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_add_ipv4_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv4_disabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_del_ipv4_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv6_enabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_add_ipv6_adj_sid(circuit, adj);
+
+	return 0;
+}
+
+static int isis_sr_adj_ipv6_disabled(struct isis_adjacency *adj)
+{
+	struct isis_circuit *circuit = adj->circuit;
+
+	if (!circuit->area->srdb.enabled)
+		return 0;
+
+	isis_sr_circuit_del_ipv6_adj_sid(circuit, adj);
+
+	return 0;
+}
+
```

Comment:
Same. IPv4 & IPv6 functions could be merge

Reply:
Agreed. I'll merge the IPv4/IPv6 Adj-SID functions to reduce code duplication. [3]